src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestStream.cs

```diff
@@ -227,5 +249,184 @@ private async Task CopyToAsyncInternal(Stream destination, CancellationToken can
                 }
             }
         }
+
+        private class AsyncPipeReader : IValueTaskSource<int>
+        {
+            private readonly PipeReader _pipeReader;
+            private readonly Action _onCompletedAction;
+
+            private IAsyncEnumerator<int> _reader;
+
+            private Memory<byte> _buffer;
+            private CancellationToken _cancellationToken;
+            private ValueTaskAwaiter<bool> _readAwaiter;
+            private short _token;
+
+            private Action<object> _continuation;
+            private object _state;
+            private ValueTaskSourceStatus _status;
+
+            public AsyncPipeReader(PipeReader pipeReader)
+            {
+                _pipeReader = pipeReader;
+                _onCompletedAction = OnCompleted;
+                _reader = ReadAsync().GetAsyncEnumerator();
+            }
+
+            private void OnCompleted()
+            {
+                var continuation = _continuation;
+                _continuation = null;
+                var state = _state;
+                _state = null;
+
+                continuation(state);
+            }
+
+            public short StartRead(Memory<byte> buffer, CancellationToken cancellationToken)
+            {
+                var token = (short)(_token + 1);
+                _token = token;
+                _buffer = buffer;
+                _cancellationToken = cancellationToken;
+
+                var task = _reader.MoveNextAsync();
```

Comment:
Is the AsyncPipeReader protected from concurrent access? If not, and an invalid concurrent read were to be attempted, would it throw a nice error and stay in a predictable state?

Reply:
Is now (ish) not protected from in anger concurrent access with interlocked

Reply:
Will throw that is